pstoremem/addr_book.go

```diff
@@ -152,55 +156,42 @@ func (mab *memoryAddrBook) UpdateAddrs(p peer.ID, oldTTL time.Duration, newTTL t
 	mab.addrmu.Lock()
```

Comment:
This PR doesn't change the locking, but since update and add operations are scoped to a peer, we might benefit from using a [striped lock](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html) vs a global lock. Any thoughts?

Reply:
I believe the tricky part would be iterating over the map. We'd need to copy the peer IDs into a separate array to do that.

Now, we *could* use a `sync.Map` but I'm worried about the memory overhead.

---

Really, I think the correct solution is a tiered store (with rotation) or something like that. That is:

* 1m tier
* 10m tier
* 1hr tier
* 1day tier
* infinity? // GC once every N days.

For each tier, we'd have an "expiring" map and a "live" map. Every time period, we'd delete the expiring map, move the live map to the expiring map, and create a new live map.

There are probably other ways to do this but this seems like the easiest solution to me.

Reply:
We could also try out `sync.Map` if it is a problem.

Reply:
> Now, we could use a sync.Map but I'm worried about the memory overhead.

But yeah, we could try it and see.

Reply:
I missed that in your message. Sorry.